### PR TITLE
Adding new section on monitoring cluster events and logs to security …

### PR DESCRIPTION
**Pull request by ghost**
…guide

This is related to PR https://github.com/openshift/openshift-docs/pull/5746

### CODE REVIEW

**Comment by tmorriso-rh** [comment]
[rev_history]
|xref:../security.adoc[Container Security Guide]
|Added the xref:../security/monitoring.adoc[Monitoring cluster events and logs] section.
%

**Comment by tmorriso-rh** [comment]
@ahardin-rh  Hi Ashley, can you peer review this PR and the revision history?

Can you also review the OpenShift v1 API and Kubernetes v1 API xrefs in the changed file? These links are broken and I am not sure why. They are formatted like the other xrefs.

Thanks!
Traci

File: security/monitoring.adoc [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+     24  user="system:serviceaccount:kube-system:pod-garbage-collector"
+     18  user="system:serviceaccount:kube-system:endpoint-controller"
+     14  user="system:serviceaccount:openshift-infra:serviceaccount-pull-secrets-controller"
+     11  user="test user"
+      4  user="test \" user"
+----
+
+More sophisticated queries generally require more sophisticated
+log analysis tools. Auditors will also need a detailed familiarity
+with the xref:../rest_api/openshift_v1.adoc#rest-api-openshift-v1:[OpenShift v1 API]
```
Comment (ahardin-rh) [inline security/monitoring.adoc]: We do not need the : after the file paths, which may be causing the issues

File: security/monitoring.adoc [hunk 2]
```diff
@@ -0,0 +1,197 @@
+[[security-monitoring]]
+= Monitoring cluster events and logs
```
Comment (ahardin-rh) [inline security/monitoring.adoc]: Monitoring Cluster Events and Logs

File: security/monitoring.adoc [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+In addition to security measures mentioned in other sections of this
+guide, the ability to monitor and audit an {product-title} cluster is an
+important part of safeguarding the cluster itself and its users against
+inappropriate usage.
+
+There are two main sources of cluster-level information that
+are useful for this purpose: events and logs.
+
+[[security-monitoring-events]]
+== Cluster events
```
Comment (ahardin-rh) [inline security/monitoring.adoc]: Cluster Events

Comment (ahardin-rh) [reply security/monitoring.adoc]: All headings on the page should be in title case

File: security/monitoring.adoc [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+configuration.
+
+Events are associated with a namespace, either the namespace of the
+resource they are related to or, for cluster events, the *_default_*
+namespace. The *_default_* namespace is likely to hold the most relevant
+events for monitoring or auditing a cluster, both *_Node_* events as well
+as resource events related to infrastructure components like the registry
+and router.
+
+The master API and `*oc*` command do not at this time provide parameters
```
Comment (ahardin-rh) [inline security/monitoring.adoc]: `oc` command

Comment (ahardin-rh) [reply security/monitoring.adoc]: https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference

File: security/monitoring.adoc [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    "uid": "origin-node-1.example.local"
+  },
+  "kind": "Event",
+  "reason": "NodeHasDiskPressure",
+  ...
+}
+----
+
+Events related to resource creation, modification, or deletion globally
+may also be good candidates for detecting misuse of the cluster. As an
```
Comment (ahardin-rh) [inline security/monitoring.adoc]: s/may/can

File: security/monitoring.adoc [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+More sophisticated queries generally require more sophisticated
+log analysis tools. Auditors will also need a detailed familiarity
+with the xref:../rest_api/openshift_v1.adoc#rest-api-openshift-v1:[OpenShift v1 API]
+and xref:../rest_api/kubernetes_v1.adoc#rest-api-kubernetes-v1:[Kubernetes v1 API]
+to aggregate request summaries from the audit log according to which
+kind of resource is involved (the *_uri_* field).
+
+xref:../install_config/master_node_configuration.adoc#master-node-config-advanced-audit[More advanced audit logging capabilities]
+are introduced with {product-title} 3.7 as a technology preview.
```
Comment (ahardin-rh) [inline security/monitoring.adoc]: Technology Preview feature

**Comment by vikram-redhat** [comment]
@gaurav-nelson - can you also review today please?

File: security/monitoring.adoc [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+:prewrap!:
+
+toc::[]
+
+[[security-monitoring-intro]]
+== Introduction
+
+In addition to security measures mentioned in other sections of this
+guide, the ability to monitor and audit an {product-title} cluster is an
+important part of safeguarding the cluster itself and its users against
```
Comment (gaurav-nelson) [inline security/monitoring.adoc]: s/the cluster itself and its/the cluster and its

File: security/monitoring.adoc [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+Cluster administrators are encouraged to familiarize themselves with the _Event_ resource
+type and review a xref:../dev_guide/events.adoc#events-reference[list of events] to
+determine which may be of interest. There are typically more potential
+event types than listed here, depending on master controller and plugin
+configuration.
+
+Events are associated with a namespace, either the namespace of the
+resource they are related to or, for cluster events, the *_default_*
+namespace. The *_default_* namespace is likely to hold the most relevant
+events for monitoring or auditing a cluster, both *_Node_* events as well
```
Comment (gaurav-nelson) [inline security/monitoring.adoc]: s/as well as/and

File: security/monitoring.adoc (hunk 4 above)
Comment (gaurav-nelson) [inline security/monitoring.adoc]: s/do not at this time provide/ do not provide

File: security/monitoring.adoc [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+as resource events related to infrastructure components like the registry
+and router.
+
+The master API and `*oc*` command do not at this time provide parameters
+to scope a listing of events to only those related to nodes. A simple
+approach would be to use grep:
+
+----
```
Comment (gaurav-nelson) [inline security/monitoring.adoc]: Maybe add [`source tags`](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#code-blocks) to all the code blocks.

File: security/monitoring.adoc [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+$ oc get event -n default | grep Node
+1h         20h         3         origin-node-1.example.local   Node      Normal    NodeHasDiskPressure   ...
+----
+
+A more flexible approach is to output the events in a form that other
+tools can then process. For example, the following example uses the `*jq*`
```
Comment (gaurav-nelson) [inline security/monitoring.adoc]: s/can then process/can process

File: security/monitoring.adoc [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+counts the number of such recent events:
+
+----
+$ oc get events --all-namespaces -o json \
+  | jq '[.items[] | select(.involvedObject.kind == "Pod" and .reason == "Pulling")] | length'
+
+4
+----
+
+Note that when a namespace is deleted, its events are deleted along with
```
Comment (gaurav-nelson) [inline security/monitoring.adoc]: Maybe we can put this info inside a `NOTE` admonition block.

File: security/monitoring.adoc [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+received and one line with the HTTP response code when it completes. For
+example, here is a record of the system administrator requesting a list
+of nodes:
+
+----
+2017-10-17T13:12:17.635085787Z AUDIT: id="410eda6b-88d4-4491-87ff-394804ca69a1" ip="192.168.122.156" method="GET" user="system:admin" groups="\"system:cluster-admins\",\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/api/v1/nodes"
+2017-10-17T13:12:17.636081056Z AUDIT: id="410eda6b-88d4-4491-87ff-394804ca69a1" response="200"
+----
+
+For this example, it may be useful to poll the log periodically using the following
```
Comment (gaurav-nelson) [inline security/monitoring.adoc]: s/periodically using/periodically by using

File: security/monitoring.adoc [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+$ tail -5000 /var/log/openshift-audit.log \
+  | grep -Po 'response="..."' \
+  | sort | uniq -c | sort -rn
+
+   3288 response="200"
+      8 response="404"
+      6 response="201"
+----
+
+Most valid requests succeed with a 200 or 201 return code. 401 and
```
Comment (gaurav-nelson) [inline security/monitoring.adoc]: s/return code/response code 

Response code is the general term used when we talk about API requests.

File: security/monitoring.adoc (hunk 6 above)
Comment (gaurav-nelson) [reply security/monitoring.adoc]: Also, it will be a good idea to move this up top in the Introduction section. For example https://docs.openshift.com/container-platform/latest/architecture/service_catalog/index.html#service-catalog-overview

**Comment by gaurav-nelson** [comment]
Thank you @tmorriso-rh, I have added few minor comments. Rest all LGTM :bar_chart:

File: security/monitoring.adoc (hunk 1 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 2 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 3 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 4 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 5 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 6 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 7 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 8 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 4 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 10 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 12 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 11 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

File: security/monitoring.adoc (hunk 13 above)
Comment (tmorriso-rh) [reply security/monitoring.adoc]: Fixed.

**Comment by vikram-redhat** [comment]
@tmorriso-rh - the tagging to @sosiouxme didn't work in your comment. There seems to be a space after the @ sign. Hopefully this tag will work :). Luke - when you get a chance, can you please review?

File: security/monitoring.adoc [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+There are two main sources of cluster-level information that
+are useful for this purpose: events and logs.
+
+[[security-monitoring-events]]
+== Cluster Events
+
+Cluster administrators are encouraged to familiarize themselves with the _Event_ resource
+type and review a xref:../dev_guide/events.adoc#events-reference[list of events] to
+determine which may be of interest. There are typically more potential
+event types than listed here, depending on master controller and plugin
```
Comment (nengard) [inline security/monitoring.adoc]: This is a bit awkward, I'd swap it to be:

"Depending on the master controller and plugin configuration, there are typically more potential event types than listed here."

File: security/monitoring.adoc [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+configuration.
+
+Events are associated with a namespace, either the namespace of the
+resource they are related to or, for cluster events, the *_default_*
+namespace. The *_default_* namespace is likely to hold the most relevant
+events for monitoring or auditing a cluster, both *_Node_* events and
+as resource events related to infrastructure components like the registry
```
Comment (nengard) [inline security/monitoring.adoc]: I'm confused by this (I wonder if a word is just missing):

" both *_Node_* events and as resource events related to infrastructure components like the registry and router."

File: security/monitoring.adoc [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+of nodes:
+
+[source, bash]
+----
+2017-10-17T13:12:17.635085787Z AUDIT: id="410eda6b-88d4-4491-87ff-394804ca69a1" ip="192.168.122.156" method="GET" user="system:admin" groups="\"system:cluster-admins\",\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/api/v1/nodes"
+2017-10-17T13:12:17.636081056Z AUDIT: id="410eda6b-88d4-4491-87ff-394804ca69a1" response="200"
+----
+
+For this example, it may be useful to poll the log periodically by using the following
+command to see if the number of failed requests:
```
Comment (nengard) [inline security/monitoring.adoc]: see if the number of failed requests .... what?

File: security/monitoring.adoc [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+$ tail -5000 /var/log/openshift-audit.log \
+  | grep -Po 'response="..."' \
+  | sort | uniq -c | sort -rn
+
+   3288 response="200"
+      8 response="404"
+      6 response="201"
+----
+
+Most valid requests succeed with a 200 or 201 response code. 401 and
```
Comment (nengard) [inline security/monitoring.adoc]: Should the response codes in this section be treated as literals and styled with monospace?

File: security/monitoring.adoc [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+403 are simple redirects, and 404 is typically a benign request for a
+resource that does not exist, but return code 400 may be of
```
Comment (nengard) [inline security/monitoring.adoc]: "but return code 400 may be of interest as it indicates a malformed request, which should not occur with most clients."

IBM Styleguide recommends not using 'may' instead use 'might'.

Also, this should probably be a new sentence instead of a 'but'.

File: security/monitoring.adoc [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+interest as it indicates a malformed request, which should not occur with
+most clients. Responses in the 500 range indicate server errors, which
+could be a result of bugs, system failures, or even malicious activity.
```
Comment (nengard) [inline security/monitoring.adoc]: IBM Styleguide states to not use 'could' - instead use 'might'

File: security/monitoring.adoc [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+If unusual numbers of error responses are found, the audit log entries
```
Comment (nengard) [inline security/monitoring.adoc]: "If unusual numbers of error responses are found,"

should be

"If an unusual number of error responses are found,"

File: security/monitoring.adoc [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+for corresponding requests can be retrieved for further investigation.
+
+It could also be useful to look for unusual numbers of requests by a
```
Comment (nengard) [inline security/monitoring.adoc]: IBM Styleguide states to not use 'could' - instead use 'might' or 'can'

File: security/monitoring.adoc [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+particular user or group. Note that the IP address of the request is typically
+a cluster host or API load balancer, and there is no record of the IP address
+behind a load balancer proxy request (however, load balancer logs could
+potentially be useful for determining request origin). The following
+example lists the top ten users by number of requests in the last 5000
```
Comment (nengard) [inline security/monitoring.adoc]: ten should be '10' per IBM styleguide (spell out numbers up to 9)

**Review by ncbaratta — CHANGES_REQUESTED** [review]
This is my first pass at a peer review on these docs. Some of my confusion might be due to lack of knowledge of the product and/or your conventions for writing docs.  I want to review again once you make updates if that's okay.

**Comment by tmorriso-rh** [comment]
@nengard Changes made. PTAL. Thanks.

File: security/monitoring.adoc [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+----
+$ oc get events --all-namespaces -o json \
+  | jq '[.items[] | select(.involvedObject.kind == "Pod" and .reason == "Pulling")] | length'
+
+4
+----
+
+[[NOTE]]
+====
+When a namespace is deleted, its events are deleted along with it. Events also expire and are removed to prevent filling up *etcd*. Events are not stored
```
Comment (nengard) [inline security/monitoring.adoc]: Consider rewording to be clearer.  Maybe something like:

"When a namespace is deleted, its events are deleted as well. Events can also expire and are deleted to prevent filling up the *etcd* file. Events are not stored as a permanent record and frequent polling is necessary to capture statistics over time."

File: security/monitoring.adoc [hunk 24]
```diff
@@ -0,0 +1,202 @@
...
+ifdef::openshift-enterprise[]
+- atomic-openshift-master-api
+- atomic-openshift-master-controllers
+- etcd
+- atomic-openshift-node
+endif::[]
+
+These logs provide a relatively verbose record of component operations
+directed more at debugging purposes than security auditing. They can be
+retrieved per host with `*journalctl*` or, in clusters with the aggregated
```
[the remark this answers is not in the document]
Comment (nengard) [reply security/monitoring.adoc]: Consider rewording in a more minimalist fashion:

"These logs are intended for debugging purposes, not security auditing."

A "log" is a "record of operations" and "relatively verbose" really doesn't add anything but words.

File: security/monitoring.adoc [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+$ tail -5000 /var/log/openshift-audit.log \
+  | grep -Po 'response="..."' \
+  | sort | uniq -c | sort -rn
+
+   3288 response="200"
+      8 response="404"
+      6 response="201"
+----
+
+Most valid requests succeed with a `200` or `201` response code. `401` and
```
Comment (nengard) [inline security/monitoring.adoc]: I feel as though this paragraph can be displayed in a more user-friendly fashion such as an easy to browse table or a bulleted list - maybe even a definition list.

File: security/monitoring.adoc [hunk 26]
```diff
@@ -0,0 +1,202 @@
...
+example, here is a record of the system administrator requesting a list
+of nodes:
+
+[source, bash]
+----
+2017-10-17T13:12:17.635085787Z AUDIT: id="410eda6b-88d4-4491-87ff-394804ca69a1" ip="192.168.122.156" method="GET" user="system:admin" groups="\"system:cluster-admins\",\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/api/v1/nodes"
+2017-10-17T13:12:17.636081056Z AUDIT: id="410eda6b-88d4-4491-87ff-394804ca69a1" response="200"
+----
+
+For this example, it might be useful to poll the log periodically by using the following
```
Comment (nengard) [inline security/monitoring.adoc]: "this" refers to what? The above example or the below example? Not that you want to use 'above' or 'below' because the IBM styleguide says not to - but we should be clearer on what we're referring to.

File: security/monitoring.adoc [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+== Cluster Events
+
+Cluster administrators are encouraged to familiarize themselves with the _Event_ resource
+type and review a xref:../dev_guide/events.adoc#events-reference[list of events] to
+determine which events might be of interest. Depending on the master controller and plugin
+configuration, there are typically more potential event types than listed here.
+
+Events are associated with a namespace, either the namespace of the
+resource they are related to or, for cluster events, the *_default_*
+namespace. The *_default_* namespace is likely to hold the most relevant
```
Comment (nengard) [inline security/monitoring.adoc]: "likely" is  vague - we should be more confident in our instructions - does it hold relevant events or not?

File: security/monitoring.adoc [hunk 28]
```diff
@@ -0,0 +1,202 @@
...
+directed more at debugging purposes than security auditing. They can be
+retrieved per host with `*journalctl*` or, in clusters with the aggregated
+logging stack deployed, in the logging _.operations_ indexes (possibly
+in the xref:../install_config/aggregate_logging.adoc#aggregated-ops[Ops
+cluster]) as a cluster administrator.
+
+[[security-monitoring-audit-log]]
+=== Master API Audit Log
+
+For security auditing purposes, there is an option to enable
```
Comment (nengard) [inline security/monitoring.adoc]: Again, this is way more wordy that we need to be.  Consider rewording - an example:

To log master API requests by users, administrators, or system components enable audit logging  +xref:../install_config/master_node_configuration.adoc#master-node-config-audit-config[for the master API]. This will create a file on each master host or, if there is no file configured, be included in the service's journal. Entries in the journal can be found by searching for "AUDIT".

File: security/monitoring.adoc [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+Most valid requests succeed with a `200` or `201` response code. `401` and
+`403` are simple redirects, and `404` is typically a benign request for a
+resource that does not exist. Return code `400` might be of
+interest as it indicates a malformed request, which should not occur with
+most clients. Responses in the `500` range indicate server errors, which
+might be a result of bugs, system failures, or even malicious activity.
+If an unusual number of error responses are found, the audit log entries
+for corresponding requests can be retrieved for further investigation.
+
+It can also be useful to look for unusual numbers of requests by a
```
Comment (nengard) [inline security/monitoring.adoc]: No need to say 'it can be useful'

File: security/monitoring.adoc [hunk 30]
```diff
@@ -0,0 +1,202 @@
...
+particular user or group. Note that the IP address of the request is typically
```
Comment (nengard) [inline security/monitoring.adoc]: The note is a bit confusing - maybe it should be styled as a note, or reworded.  I think you're saying that searching for IPs isn't always a foolproof way.

File: security/monitoring.adoc [hunk 31]
```diff
@@ -0,0 +1,202 @@
...
+     66  user="system:anonymous"
+     32  user="system:serviceaccount:kube-system:cronjob-controller"
+     24  user="system:serviceaccount:kube-system:pod-garbage-collector"
+     18  user="system:serviceaccount:kube-system:endpoint-controller"
+     14  user="system:serviceaccount:openshift-infra:serviceaccount-pull-secrets-controller"
+     11  user="test user"
+      4  user="test \" user"
+----
+
+More sophisticated queries generally require more sophisticated
```
Comment (nengard) [inline security/monitoring.adoc]: Is it necessary to say this? If so - think of a synonym for sophisticated for one of the two uses of the word.

File: security/monitoring.adoc [hunk 32]
```diff
@@ -0,0 +1,202 @@
...
+log analysis tools. Auditors will also need a detailed familiarity
+with the xref:../rest_api/openshift_v1.adoc#rest-api-openshift-v1[OpenShift v1 API]
+and xref:../rest_api/kubernetes_v1.adoc#rest-api-kubernetes-v1[Kubernetes v1 API]
+to aggregate request summaries from the audit log according to which
+kind of resource is involved (the *_uri_* field).
```
Comment (nengard) [inline security/monitoring.adoc]: We should never say 'kind of' - it's very wishy washy.  We are the experts and should give clear instructions.

Comment (nengard) [inline security/monitoring.adoc]: If the field is labeled 'uri' then it should only be bold - not italics. Otherwise, remove bold and make just italics. We don't need both.

File: security/monitoring.adoc [hunk 33]
```diff
@@ -0,0 +1,202 @@
...
+
+xref:../install_config/master_node_configuration.adoc#master-node-config-advanced-audit[More advanced audit logging capabilities]
+are introduced with {product-title} 3.7 as a xref:..https://access.redhat.com/support/offerings/techpreview[Technology Preview] feature.
+This feature enables providing an audit policy file to control which
+requests are logged and the level of detail to log. Advanced audit
+log entries provide more detail in JSON format and might be logged via a
```
Comment (nengard) [inline security/monitoring.adoc]: Don't use 'might' per the IBM styleguide.

**Review by ncbaratta — CHANGES_REQUESTED** [review]

File: security/monitoring.adoc (hunk 32 above)
Comment (sosiouxme) [reply security/monitoring.adoc]: "kind" is a technical term here (as is "resource"). The type of resource, not "resource-like"

File: security/monitoring.adoc (hunk 27 above)
Comment (sosiouxme) [reply security/monitoring.adoc]: Depends on what you're interested in / what you consider relevant. This namespace holds events relevant to the entire cluster. Other namespaces may also have events that are of interest to the user, depending on what they're interested in. Perhaps reword not to assert which are "most" relevant.

**Comment by tmorriso-rh** [comment]
@nengard Changes made. PTAL. Thanks.

File: security/monitoring.adoc [hunk 34]
```diff
@@ -0,0 +1,196 @@
...
+----
+$ oc get events --all-namespaces -o json \
+  | jq '[.items[] | select(.involvedObject.kind == "Pod" and .reason == "Pulling")] | length'
+
+4
+----
+
+[NOTE]
+====
+When a namespace is deleted, its events are deleted as well. Events can also expire and are deleted to prevent filling up the *etcd* file. Events are
```
Comment (sosiouxme) [inline security/monitoring.adoc]: etcd is not a file, it's a database. Prefer to say "to prevent filling etcd storage."

File: security/monitoring.adoc [hunk 35]
```diff
@@ -0,0 +1,196 @@
...
+not stored as a permanent record and frequent polling is necessary to capture statistics over time.
+====
+
+[[security-monitoring-logs]]
+== Cluster Logs
+This section describes the types of monitoring logs produced on the cluster.
```
Comment (sosiouxme) [inline security/monitoring.adoc]: "monitoring logs" doesn't really mean anything. The cluster hosts just produce... logs. If you need an adjective I would use "operational logs".

File: security/monitoring.adoc [hunk 36]
```diff
@@ -0,0 +1,196 @@
...
+- origin-node
+endif::[]
+ifdef::openshift-enterprise[]
+- atomic-openshift-master-api
+- atomic-openshift-master-controllers
+- etcd
+- atomic-openshift-node
+endif::[]
+
+These logs are intended for debugging purposes, not security auditing. They can be
```
Comment (sosiouxme) [inline security/monitoring.adoc]: This sounds too negative. I wouldn't say that they're not intended for security auditing *at all*. That's just not their focus.

Maybe:
"These logs are intended more for debugging purposes than security auditing."

File: security/monitoring.adoc (hunk 25 above)
Comment (sosiouxme) [reply security/monitoring.adoc]: Maybe, but this is far from a complete list of possible HTTP response codes, and we wouldn't want to give that impression. It's just the highlights.

File: security/monitoring.adoc [hunk 37]
```diff
@@ -0,0 +1,196 @@
...
+of nodes:
+
+[source, bash]
+----
+2017-10-17T13:12:17.635085787Z AUDIT: id="410eda6b-88d4-4491-87ff-394804ca69a1" ip="192.168.122.156" method="GET" user="system:admin" groups="\"system:cluster-admins\",\"system:authenticated\"" as="<self>" asgroups="<lookup>" namespace="<none>" uri="/api/v1/nodes"
+2017-10-17T13:12:17.636081056Z AUDIT: id="410eda6b-88d4-4491-87ff-394804ca69a1" response="200"
+----
+
+When requesting the list, it can be useful to poll the log periodically by using the following
+command to see if the number of requests failed:
```
Comment (sosiouxme) [inline security/monitoring.adoc]: This doesn't make sense. Requesting what list? To see if the number what?

The example command shows the response codes seen in the last 5000 lines of the log and the number of responses that had each code. "Failure" isn't even clearly defined here. Auditors will have their own opinions about what is of concern.

Suggest: "It could be useful to poll the log periodically for the number of recent requests per response code, as in the following example:"

**Review by sosiouxme — CHANGES_REQUESTED** [review]

**Comment by tmorriso-rh** [comment]
@sosiouxme @nengard Made additional changes. PTAL.

File: security/monitoring.adoc [hunk 38]
```diff
@@ -0,0 +1,194 @@
...
+----
+$ tail -5000 /var/log/openshift-audit.log \
+  | grep -Po 'response="..."' \
+  | sort | uniq -c | sort -rn
+
+   3288 response="200"
+      8 response="404"
+      6 response="201"
+----
+The following list describes the response codes in more detail:
```
Comment (sosiouxme) [inline security/monitoring.adoc]: I suggest saying "some of the common response codes" to avoid implying this list is complete.

File: security/monitoring.adoc [hunk 39]
```diff
@@ -0,0 +1,194 @@
...
+
+* `200` or `201` response codes indicate a successful request.
+* `400` response codes may be of interest as they indicate a malformed request, which should not occur with most clients.
+* `401` and `403` response codes are simple redirects.
+* `404` response codes are typically benign requests for a resource that does not exist.
+* `500` response codes indicate server errors, which can be a result of bugs, system failures, or even malicious activity.
```
Comment (sosiouxme) [inline security/monitoring.adoc]: It is not only `500` but any code in the 500-599 range that indicates some kind of server error. Sometimes written as `5XX`

File: security/monitoring.adoc [hunk 40]
```diff
@@ -0,0 +1,194 @@
...
+
+If an unusual number of error responses are found, the audit log entries for corresponding requests can be retrieved for
+further investigation. It can be useful to look for unusual numbers of requests by a particular user or group.
```
Comment (sosiouxme) [inline security/monitoring.adoc]: These two sentences are not related; having finished talking about return codes, we want to go on to talk about numbers of requests collated by user/group. The NOTE is in the middle below because admins would probably also like to be able to collate by source IP, but you usually can't. Then it goes on to describe how to do that first thing.

**Review by sosiouxme — COMMENTED** [review]
Nothing critical but a few improvements to mention.

**Comment by tmorriso-rh** [comment]
@sosiouxme Changes made. PTAL.

File: security/monitoring.adoc [hunk 41]
```diff
@@ -0,0 +1,197 @@
...
+      6 response="201"
+----
+The following list describes some of the response codes in more detail:
+
+* `200` or `201` response codes indicate a successful request.
+* `400` response codes may be of interest as they indicate a malformed request, which should not occur with most clients.
+* `401` and `403` response codes are simple redirects.
+* `404` response codes are typically benign requests for a resource that does not exist.
+* `500` - `599` response codes indicate server errors, which can be a result of bugs, system failures, or even malicious activity. These
+response codes can also be written as `5XX`.
```
Comment (sosiouxme) [inline security/monitoring.adoc]: Ah, I should have been clearer. They won't literally be `5XX`. That's just how the range 500-599 is sometimes represented in discussion.

**Review by sosiouxme — APPROVED** [review]
Otherwise, LGTM!

**Comment by tmorriso-rh** [comment]
Thanks @sosiouxme. I removed the `5XX` line. Please review.

File: security/monitoring.adoc [hunk 42]
```diff
@@ -0,0 +1,191 @@
...
+   3288 response="200"
+      8 response="404"
+      6 response="201"
+----
+
+The following list describes some of the response codes in more detail:
+
+* `200` or `201` response codes indicate a successful request.
+* `400` response codes may be of interest as they indicate a malformed request, which should not occur with most clients.
+* `401` and `403` response codes are simple redirects.
```
Comment (sosiouxme) [inline security/monitoring.adoc]: Wow, I don't know where my brain went on this. The return codes I meant here should be 301 and 302. But there's probably no real reason to mention these at all (I must have been thinking the 4XX range would be confusing with some codes "normal" and some "suspicious"). So I'd say just delete this line. My mistake!

**Review by sosiouxme — APPROVED** [review]
Otherwise - looks good, ship it!

**Comment by sosiouxme** [comment]
:tada:

**Comment by vikram-redhat** [comment]
@tmorriso-rh - this PR hasn't been picked to the 3.7 staging branch, only to the 3.6 branch. Can you make sure to pick it to 3.7 staging branch?

**Comment by tmorriso-rh** [comment]
@vikram-redhat Hi Vikram, this is the cherry-picked PR that was opened for this PR:

https://github.com/openshift/openshift-docs/pull/6385

Do I need to re-do the cherry pick? I just want to make sure I am not aggravating the issue. Thanks.

**Comment by vikram-redhat** [comment]
Hey @tmorriso-rh - https://github.com/openshift/openshift-docs/pull/6385 seems to be for another issue: "Adding new section on searching for images..". This PR is for new section on monitoring cluster events  and logs to security. They are different PRs.Yes, this PR still needs to be CP to 3.7 stage unless you have created a PR but not pushed it.

**Comment by sosiouxme** [comment]
If this PR *were* picked to 3.6 docs it would need a minor modification not
to mention the advanced audit log which is tech preview 3.7.

**Comment by vikram-redhat** [comment]
Thanks @sosiouxme. @tmorriso-rh is update the 3.6-stage branch to take that reference out.

**Comment by vikram-redhat** [comment]
Hey @tmorriso-rh - just FYI, the rev history linked to a wrong file: security.adoc. It should have been security/index.adoc (or similar). I have fixed it for now.
